lib/THC/THCTensorRandom.cu

```diff
@@ -359,4 +359,405 @@ THC_API void THCudaTensor_cauchy(THCState* state, THCudaTensor *self_, double me
+                                    float* dest,
+                                    long distributions,
+                                    int categories,
+                                    float* origDist,
+                                    float* normDistPrefixSum) {
+  // Binary search is by its nature warp divergent, so all threads
+  // in a warp compute the save value. It's important that all threads
+  // in a block particiapte in the curand_uniform call to ensure that
+  // the random number generator state is fully updated.
+
+  // The block and warp determines the distribution for which we
+  // generate a point
+  for (long curDistBase = blockIdx.x * blockDim.y;
+       curDistBase < distributions;
+       curDistBase += gridDim.x * blockDim.y) {
+    long curDist = curDistBase + threadIdx.y;
+    float r = curand_uniform(&state[blockIdx.x]);
+
+    if (threadIdx.x == 0 && curDist < distributions) {
+      // Find the bucket that a uniform sample lies in
+      int choice = binarySearchForMultinomial(
+        normDistPrefixSum + curDist * categories,
+        categories,
+        r);
+
+      // Torch indices are 1-based
+      dest[curDist * totalSamples + sample] = (float) choice + 1.0f;
+
+      // Without replacement, so update the original probability so it
+      // is not considered a second time
+      origDist[curDist * categories + choice] = 0.0f;
+    }
+  }
+}
+
+THC_API void THCudaTensor_multinomial(struct THCState *state,
+                                      THCudaTensor *self,
+                                      THCudaTensor *prob_dist,
+                                      int n_sample,
+                                      int with_replacement)
+{
+  THAssert(THCudaTensor_checkGPU(state, 2, self, prob_dist));
+  if (state->rngState->current_gen == NULL) {
+    THError("Random number generators have not been initialized.");
+  }
+
+  int inputSize = THCudaTensor_nDimension(state, prob_dist);
+  THArgCheck(inputSize > 0 && inputSize <= 2, 2,
+             "prob_dist must be 1 or 2 dim");
+
+  // Categories are in the innermost dimension
+  long numDist =
+    inputSize == 1 ? 1 : THCudaTensor_size(state, prob_dist, 0);
+  long numCategoriesLong =
+    inputSize == 1 ? THCudaTensor_size(state, prob_dist, 0) :
+    THCudaTensor_size(state, prob_dist, 1);
+
+  // Since the index tensor is float, numCategories cannot exceed max=
+  // float integer precision
+  THArgCheck(numCategoriesLong <= FLOAT32_MAX_CONSECUTIVE_INT, 2,
+             "number of categories cannot exceed 2^24");
+  int numCategories = (int) numCategoriesLong;
+
+  THArgCheck(n_sample > 0, 3, "cannot sample n_sample < 0 samples");
+
+  if (!with_replacement) {
+    THArgCheck(n_sample <= numCategories, 2,
+               "cannot sample n_sample > prob_dist:size(1) samples without "
+               "replacement");
+  }
+
+  // Restructure data for 2d
+  if (inputSize == 1) {
+    THCudaTensor_resize2d(state, prob_dist, 1, numCategories);
```

Comment:
This resize operation only works as expected if prob_dist is contiguous, because resize ignores the tensor's stride. So you should make it contiguous before you do this (you make it contiguous later anyway).

The below one is fine, because it's not an input.